{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# BTM Data science notebook\n",
    "\n",
    "In this session we will use examples from https://github.com/pycam/python-data-science condensed down to a 1 hour session, to get familiar with pandas and classes. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Pandas \n",
    "Pandas is a data analysis module that allows you to use python with tables and data objects, much like data.frames in R.\n",
    "\n",
    "Lets get started by opening a table and look at some data using pandas. We will also look at how to do the same taks in pyhton and see how much simpler pandas can be sometimes."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reading a table without pandas "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BRCA2 13 32889611 32973805\n",
      "TNFAIP3 6 138188351 138204449\n",
      "TCF7 5 133450402 133487556\n"
     ]
    }
   ],
   "source": [
    "import csv\n",
    "data = []\n",
    "with open(\"data/genes.txt\") as f:\n",
    "    reader = csv.DictReader(f, delimiter = \"\\t\")\n",
    "    for row in reader:\n",
    "        data.append(row)\n",
    "# look at the data\n",
    "for d in data:\n",
    "    print(d['gene'], d['chrom'], d['start'], d['end'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Loading the same data using pandas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>13</td>\n",
       "      <td>32889611</td>\n",
       "      <td>32973805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TNFAIP3</td>\n",
       "      <td>6</td>\n",
       "      <td>138188351</td>\n",
       "      <td>138204449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCF7</td>\n",
       "      <td>5</td>\n",
       "      <td>133450402</td>\n",
       "      <td>133487556</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene  chrom      start        end\n",
       "0    BRCA2     13   32889611   32973805\n",
       "1  TNFAIP3      6  138188351  138204449\n",
       "2     TCF7      5  133450402  133487556"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "df = pd.read_csv(\"data/genes.txt\", sep = \"\\t\")\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As you can see, using pandas you can be quicker in writing code that handles data tables, but you also do not have total control over what happens, so sometimes using pandas might not what you need. This you will have to decide every times new."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exploring the data\n",
    "When you firts open a data file in jupyter lab, you might want to check that the data was correctly loaded. A good way to do this is to actually lookt at the data. Pandas allows you to do this. Lets check it out."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load pandas, although it should already be loaded into this kernel\n",
    "import pandas as pd\n",
    "# open a dataset\n",
    "df = pd.read_csv('data/GRCm38.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gbid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000073.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>CDS</td>\n",
       "      <td>49420174</td>\n",
       "      <td>49420340</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>ID=cds30916;Parent=rna43633;Dbxref=GeneID:7828...</td>\n",
       "      <td>XP_006541352.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>CDS</td>\n",
       "      <td>64607740</td>\n",
       "      <td>64608137</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ID=cds1286;Parent=rna1815;Dbxref=GeneID:67099,...</td>\n",
       "      <td>XP_006496267.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000071.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>24544129</td>\n",
       "      <td>24544373</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id325971;Parent=rna28488;Dbxref=GeneID:2310...</td>\n",
       "      <td>NM_177676.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000074.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>71655663</td>\n",
       "      <td>71655741</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id568467;Parent=rna50916;Dbxref=GeneID:3820...</td>\n",
       "      <td>XM_017312883.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000068.7</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>CDS</td>\n",
       "      <td>135932249</td>\n",
       "      <td>135932302</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>ID=cds9263;Parent=rna13273;Dbxref=GeneID:18798...</td>\n",
       "      <td>XP_006499009.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         seqid      source  type      start        end score strand phase  \\\n",
       "0  NC_000073.6      Gnomon   CDS   49420174   49420340     .      +     1   \n",
       "1  NC_000067.6      Gnomon   CDS   64607740   64608137     .      -     2   \n",
       "2  NC_000071.6  BestRefSeq  exon   24544129   24544373     .      -     .   \n",
       "3  NC_000074.6      Gnomon  exon   71655663   71655741     .      -     .   \n",
       "4  NC_000068.7      Gnomon   CDS  135932249  135932302     .      +     1   \n",
       "\n",
       "                                          attributes            gbid  \n",
       "0  ID=cds30916;Parent=rna43633;Dbxref=GeneID:7828...  XP_006541352.1  \n",
       "1  ID=cds1286;Parent=rna1815;Dbxref=GeneID:67099,...  XP_006496267.1  \n",
       "2  ID=id325971;Parent=rna28488;Dbxref=GeneID:2310...     NM_177676.6  \n",
       "3  ID=id568467;Parent=rna50916;Dbxref=GeneID:3820...  XM_017312883.1  \n",
       "4  ID=cds9263;Parent=rna13273;Dbxref=GeneID:18798...  XP_006499009.1  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# lets look at the first few rows\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "seqid         object\n",
       "source        object\n",
       "type          object\n",
       "start          int64\n",
       "end            int64\n",
       "score         object\n",
       "strand        object\n",
       "phase         object\n",
       "attributes    object\n",
       "gbid          object\n",
       "dtype: object"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# lets check out the column types\n",
    "df.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(29998, 10)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# how large is the table (rows, columns)\n",
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['seqid', 'source', 'type', 'start', 'end', 'score', 'strand',\n",
       "       'phase', 'attributes', 'gbid'], dtype=object)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# what are the column names\n",
    "df.columns.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>2.999800e+04</td>\n",
       "      <td>2.999800e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>7.446804e+07</td>\n",
       "      <td>7.448115e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>4.490600e+07</td>\n",
       "      <td>4.490355e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>2.010000e+03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>3.497905e+07</td>\n",
       "      <td>3.499334e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>7.334594e+07</td>\n",
       "      <td>7.336266e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>1.088366e+08</td>\n",
       "      <td>1.088385e+08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>1.951710e+08</td>\n",
       "      <td>1.951712e+08</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              start           end\n",
       "count  2.999800e+04  2.999800e+04\n",
       "mean   7.446804e+07  7.448115e+07\n",
       "std    4.490600e+07  4.490355e+07\n",
       "min    1.000000e+00  2.010000e+03\n",
       "25%    3.497905e+07  3.499334e+07\n",
       "50%    7.334594e+07  7.336266e+07\n",
       "75%    1.088366e+08  1.088385e+08\n",
       "max    1.951710e+08  1.951712e+08"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# for numeric data pandas can give you some stats, so you can check for basic errors\n",
    "df.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gbid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12608</th>\n",
       "      <td>NT_039697.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>1</td>\n",
       "      <td>46440</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1205979;Dbxref=taxon:10090;Name=19;chromo...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1280</th>\n",
       "      <td>NT_039198.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>1</td>\n",
       "      <td>272653</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1207875;Dbxref=taxon:10090;Name=1;chromos...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23397</th>\n",
       "      <td>NC_000076.6</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>1</td>\n",
       "      <td>130694993</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id670068;Dbxref=taxon:10090;Name=10;chromos...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22001</th>\n",
       "      <td>NT_187018.1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>1896</td>\n",
       "      <td>2010</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1225737;Parent=rna112018;Dbxref=GeneID:20...</td>\n",
       "      <td>NM_009199.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20978</th>\n",
       "      <td>NT_039194.5</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>3153</td>\n",
       "      <td>3341</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds77279;Parent=rna109657;Dbxref=GeneID:263...</td>\n",
       "      <td>NP_036070.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             seqid      source    type  start        end score strand phase  \\\n",
       "12608  NT_039697.1      RefSeq  region      1      46440     .      +     .   \n",
       "1280   NT_039198.1      RefSeq  region      1     272653     .      +     .   \n",
       "23397  NC_000076.6      RefSeq  region      1  130694993     .      +     .   \n",
       "22001  NT_187018.1  BestRefSeq    exon   1896       2010     .      +     .   \n",
       "20978  NT_039194.5  BestRefSeq     CDS   3153       3341     .      +     0   \n",
       "\n",
       "                                              attributes         gbid  \n",
       "12608  ID=id1205979;Dbxref=taxon:10090;Name=19;chromo...          NaN  \n",
       "1280   ID=id1207875;Dbxref=taxon:10090;Name=1;chromos...          NaN  \n",
       "23397  ID=id670068;Dbxref=taxon:10090;Name=10;chromos...          NaN  \n",
       "22001  ID=id1225737;Parent=rna112018;Dbxref=GeneID:20...  NM_009199.2  \n",
       "20978  ID=cds77279;Parent=rna109657;Dbxref=GeneID:263...  NP_036070.2  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# sorting the table based on column\n",
    "df = df.sort_values(['start', 'end', 'strand'])\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gbid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12608</th>\n",
       "      <td>NT_039697.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>1</td>\n",
       "      <td>46440</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1205979;Dbxref=taxon:10090;Name=19;chromo...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1280</th>\n",
       "      <td>NT_039198.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>1</td>\n",
       "      <td>272653</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1207875;Dbxref=taxon:10090;Name=1;chromos...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             seqid  source    type  start     end score strand phase  \\\n",
       "12608  NT_039697.1  RefSeq  region      1   46440     .      +     .   \n",
       "1280   NT_039198.1  RefSeq  region      1  272653     .      +     .   \n",
       "\n",
       "                                              attributes gbid  \n",
       "12608  ID=id1205979;Dbxref=taxon:10090;Name=19;chromo...  NaN  \n",
       "1280   ID=id1207875;Dbxref=taxon:10090;Name=1;chromos...  NaN  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get the first two rows\n",
    "df[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "12608    NT_039697.1\n",
       "1280     NT_039198.1\n",
       "Name: seqid, dtype: object"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get only the first two seqids\n",
    "df[:2]['seqid']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gbid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20978</th>\n",
       "      <td>NT_039194.5</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>3153</td>\n",
       "      <td>3341</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds77279;Parent=rna109657;Dbxref=GeneID:263...</td>\n",
       "      <td>NP_036070.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1103</th>\n",
       "      <td>NT_039697.1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>5250</td>\n",
       "      <td>5428</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>ID=cds76922;Parent=rna109385;Dbxref=GeneID:104...</td>\n",
       "      <td>NP_659117.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20123</th>\n",
       "      <td>NT_039697.1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>6052</td>\n",
       "      <td>6087</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ID=cds76922;Parent=rna109385;Dbxref=GeneID:104...</td>\n",
       "      <td>NP_659117.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29435</th>\n",
       "      <td>NT_187026.1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>6426</td>\n",
       "      <td>6713</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds77987;Parent=rna110684;Dbxref=GeneID:216...</td>\n",
       "      <td>NP_001278048.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18061</th>\n",
       "      <td>NT_187026.1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>9062</td>\n",
       "      <td>9150</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>ID=cds77985;Parent=rna110682;Dbxref=GeneID:216...</td>\n",
       "      <td>NP_001013432.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             seqid      source type  start   end score strand phase  \\\n",
       "20978  NT_039194.5  BestRefSeq  CDS   3153  3341     .      +     0   \n",
       "1103   NT_039697.1  BestRefSeq  CDS   5250  5428     .      -     1   \n",
       "20123  NT_039697.1  BestRefSeq  CDS   6052  6087     .      -     2   \n",
       "29435  NT_187026.1  BestRefSeq  CDS   6426  6713     .      +     0   \n",
       "18061  NT_187026.1  BestRefSeq  CDS   9062  9150     .      +     2   \n",
       "\n",
       "                                              attributes            gbid  \n",
       "20978  ID=cds77279;Parent=rna109657;Dbxref=GeneID:263...     NP_036070.2  \n",
       "1103   ID=cds76922;Parent=rna109385;Dbxref=GeneID:104...     NP_659117.2  \n",
       "20123  ID=cds76922;Parent=rna109385;Dbxref=GeneID:104...     NP_659117.2  \n",
       "29435  ID=cds77987;Parent=rna110684;Dbxref=GeneID:216...  NP_001278048.1  \n",
       "18061  ID=cds77985;Parent=rna110682;Dbxref=GeneID:216...  NP_001013432.1  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filter by value and show top hits\n",
    "df[df.type == \"CDS\"].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### selecting using iloc and loc\n",
    "We see that `.iloc[0]` will return the first row in the DataFrame, while `.loc[0]` will return the row with index 0:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "seqid                                               NT_039697.1\n",
       "source                                                   RefSeq\n",
       "type                                                     region\n",
       "start                                                         1\n",
       "end                                                       46440\n",
       "score                                                         .\n",
       "strand                                                        +\n",
       "phase                                                         .\n",
       "attributes    ID=id1205979;Dbxref=taxon:10090;Name=19;chromo...\n",
       "gbid                                                        NaN\n",
       "Name: 12608, dtype: object"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.iloc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "seqid                                               NC_000073.6\n",
       "source                                                   Gnomon\n",
       "type                                                        CDS\n",
       "start                                                  49420174\n",
       "end                                                    49420340\n",
       "score                                                         .\n",
       "strand                                                        +\n",
       "phase                                                         1\n",
       "attributes    ID=cds30916;Parent=rna43633;Dbxref=GeneID:7828...\n",
       "gbid                                             XP_006541352.1\n",
       "Name: 0, dtype: object"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functions and Classes\n",
    "When we write code it is good practice to use functions and classes. They provide structure for your code and make it not only understandable but also reusable. \n",
    "\n",
    "## Functions\n",
    "Functions are short generalized code structures that can be called with an input then do some computation and return an output. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def this_is_the_function_name(input_argument1, input_argument2):\n",
    "\n",
    "    # The body of the function is indented\n",
    "    # This function prints the two arguments to screen\n",
    "    print('The function arguments are:', input_argument1, \"and\", input_argument2, '\\n(this is done inside the function!)')\n",
    "\n",
    "    # And returns their product\n",
    "    return input_argument1 * input_argument2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The function arguments are: 100 and 2 \n",
      "(this is done inside the function!)\n",
      "The result of the computation is: 200\n"
     ]
    }
   ],
   "source": [
    "# call the function\n",
    "r = this_is_the_function_name(100, 2)\n",
    "print(f\"The result of the computation is: {r}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Classes\n",
    "Classes are larger structures that are defined to group functions around a root case together. Commonly they provide functions for a certain data object. So you could for example imagine a gff3 class, which provides functionality like loading, sorting and filtering of a tsv file. Here is an example how this could look like:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "class tsvClass():\n",
    "    def __init__(self, path, autoload = False):\n",
    "        '''\n",
    "        When initialized the class will always execute this __init__ function.\n",
    "        So whatever we do here, will always be done.\n",
    "        '''\n",
    "        self.path = path\n",
    "        if autoload:\n",
    "            self.load()\n",
    "    \n",
    "    def load(self):\n",
    "        '''\n",
    "        In this function we load the gff3 file by using pandas\n",
    "        '''\n",
    "        self.df = pd.read_csv(self.path, sep = '\\t')\n",
    "    \n",
    "    def filter(self, column = \"type\", match = \"CDS\"):\n",
    "        '''\n",
    "        This filter function takes two arguments and returns a subset\n",
    "        based on these\n",
    "        '''\n",
    "        return self.df[self.df[column] == match]\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gbid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000073.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>CDS</td>\n",
       "      <td>49420174</td>\n",
       "      <td>49420340</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>ID=cds30916;Parent=rna43633;Dbxref=GeneID:7828...</td>\n",
       "      <td>XP_006541352.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>CDS</td>\n",
       "      <td>64607740</td>\n",
       "      <td>64608137</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ID=cds1286;Parent=rna1815;Dbxref=GeneID:67099,...</td>\n",
       "      <td>XP_006496267.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000071.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>24544129</td>\n",
       "      <td>24544373</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id325971;Parent=rna28488;Dbxref=GeneID:2310...</td>\n",
       "      <td>NM_177676.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000074.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>71655663</td>\n",
       "      <td>71655741</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id568467;Parent=rna50916;Dbxref=GeneID:3820...</td>\n",
       "      <td>XM_017312883.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000068.7</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>CDS</td>\n",
       "      <td>135932249</td>\n",
       "      <td>135932302</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>ID=cds9263;Parent=rna13273;Dbxref=GeneID:18798...</td>\n",
       "      <td>XP_006499009.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         seqid      source  type      start        end score strand phase  \\\n",
       "0  NC_000073.6      Gnomon   CDS   49420174   49420340     .      +     1   \n",
       "1  NC_000067.6      Gnomon   CDS   64607740   64608137     .      -     2   \n",
       "2  NC_000071.6  BestRefSeq  exon   24544129   24544373     .      -     .   \n",
       "3  NC_000074.6      Gnomon  exon   71655663   71655741     .      -     .   \n",
       "4  NC_000068.7      Gnomon   CDS  135932249  135932302     .      +     1   \n",
       "\n",
       "                                          attributes            gbid  \n",
       "0  ID=cds30916;Parent=rna43633;Dbxref=GeneID:7828...  XP_006541352.1  \n",
       "1  ID=cds1286;Parent=rna1815;Dbxref=GeneID:67099,...  XP_006496267.1  \n",
       "2  ID=id325971;Parent=rna28488;Dbxref=GeneID:2310...     NM_177676.6  \n",
       "3  ID=id568467;Parent=rna50916;Dbxref=GeneID:3820...  XM_017312883.1  \n",
       "4  ID=cds9263;Parent=rna13273;Dbxref=GeneID:18798...  XP_006499009.1  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GRCm38 = tsvClass('data/GRCm38.tsv')\n",
    "GRCm38.load()\n",
    "GRCm38.df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gbid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000071.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>24544129</td>\n",
       "      <td>24544373</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id325971;Parent=rna28488;Dbxref=GeneID:2310...</td>\n",
       "      <td>NM_177676.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000074.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>71655663</td>\n",
       "      <td>71655741</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id568467;Parent=rna50916;Dbxref=GeneID:3820...</td>\n",
       "      <td>XM_017312883.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NC_000073.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>118329486</td>\n",
       "      <td>118329612</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id520282;Parent=rna46416;Dbxref=GeneID:1026...</td>\n",
       "      <td>XR_378454.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NC_000076.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>5175722</td>\n",
       "      <td>5175928</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id673061;Parent=rna60010;Dbxref=GeneID:6400...</td>\n",
       "      <td>XM_017314047.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NC_000083.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>24585923</td>\n",
       "      <td>24586066</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1054336;Parent=rna94586;Dbxref=GeneID:187...</td>\n",
       "      <td>XM_017317334.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         seqid      source  type      start        end score strand phase  \\\n",
       "2  NC_000071.6  BestRefSeq  exon   24544129   24544373     .      -     .   \n",
       "3  NC_000074.6      Gnomon  exon   71655663   71655741     .      -     .   \n",
       "7  NC_000073.6      Gnomon  exon  118329486  118329612     .      +     .   \n",
       "8  NC_000076.6      Gnomon  exon    5175722    5175928     .      -     .   \n",
       "9  NC_000083.6      Gnomon  exon   24585923   24586066     .      +     .   \n",
       "\n",
       "                                          attributes            gbid  \n",
       "2  ID=id325971;Parent=rna28488;Dbxref=GeneID:2310...     NM_177676.6  \n",
       "3  ID=id568467;Parent=rna50916;Dbxref=GeneID:3820...  XM_017312883.1  \n",
       "7  ID=id520282;Parent=rna46416;Dbxref=GeneID:1026...     XR_378454.2  \n",
       "8  ID=id673061;Parent=rna60010;Dbxref=GeneID:6400...  XM_017314047.1  \n",
       "9  ID=id1054336;Parent=rna94586;Dbxref=GeneID:187...  XM_017317334.1  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# now we can use our filter function to reduce to exons\n",
    "GRCm38.filter(match = \"exon\").head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gbid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000073.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>CDS</td>\n",
       "      <td>49420174</td>\n",
       "      <td>49420340</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>ID=cds30916;Parent=rna43633;Dbxref=GeneID:7828...</td>\n",
       "      <td>XP_006541352.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000068.7</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>CDS</td>\n",
       "      <td>135932249</td>\n",
       "      <td>135932302</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>ID=cds9263;Parent=rna13273;Dbxref=GeneID:18798...</td>\n",
       "      <td>XP_006499009.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>87192486</td>\n",
       "      <td>87192641</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>ID=cds2054;Parent=rna2870;Dbxref=CCDS:CCDS1512...</td>\n",
       "      <td>NP_067611.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NC_000073.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>118329486</td>\n",
       "      <td>118329612</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id520282;Parent=rna46416;Dbxref=GeneID:1026...</td>\n",
       "      <td>XR_378454.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NC_000083.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>24585923</td>\n",
       "      <td>24586066</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1054336;Parent=rna94586;Dbxref=GeneID:187...</td>\n",
       "      <td>XM_017317334.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         seqid      source  type      start        end score strand phase  \\\n",
       "0  NC_000073.6      Gnomon   CDS   49420174   49420340     .      +     1   \n",
       "4  NC_000068.7      Gnomon   CDS  135932249  135932302     .      +     1   \n",
       "6  NC_000067.6  BestRefSeq   CDS   87192486   87192641     .      +     1   \n",
       "7  NC_000073.6      Gnomon  exon  118329486  118329612     .      +     .   \n",
       "9  NC_000083.6      Gnomon  exon   24585923   24586066     .      +     .   \n",
       "\n",
       "                                          attributes            gbid  \n",
       "0  ID=cds30916;Parent=rna43633;Dbxref=GeneID:7828...  XP_006541352.1  \n",
       "4  ID=cds9263;Parent=rna13273;Dbxref=GeneID:18798...  XP_006499009.1  \n",
       "6  ID=cds2054;Parent=rna2870;Dbxref=CCDS:CCDS1512...     NP_067611.2  \n",
       "7  ID=id520282;Parent=rna46416;Dbxref=GeneID:1026...     XR_378454.2  \n",
       "9  ID=id1054336;Parent=rna94586;Dbxref=GeneID:187...  XM_017317334.1  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# or we filter on the strand\n",
    "GRCm38.filter(column = \"strand\", match = \"+\").head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And because this is a class we now can go back to the gff object and just ask it what file it was from"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'data/GRCm38.tsv'"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GRCm38.path"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I hope you now got a glimpse in how to write functions, you will need time and practize to write classes that are not just spagetti code in disguise but it is worth it.\n",
    "\n",
    "Often a script can be split into functions and then put into a class which makes it more understandable and you will find bugs much quicker.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Project \n",
    "We want to query esemble with coding regions from different species. For this in the data folder you find the gff files of mouse, human and zebra fish. We will load the gff3 files and sample 100 coding sequences (CDS) for each and compare GC content between species by plotting the result. \n",
    "\n",
    "You find all needed code here you just need to make it work. If you have the time, make it nice."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests, sys\n",
    "import random\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_genomic_content(chrom, start, end, species = 'human'):\n",
    "    server = \"https://rest.ensembl.org\"\n",
    "    ext = f\"/sequence/region/{species}/{chrom}:{start}..{end}:1?\"\n",
    "\n",
    "    r = requests.get(server+ext, headers={ \"Content-Type\" : \"text/plain\"})\n",
    "\n",
    "    if not r.ok:\n",
    "      r.raise_for_status()\n",
    "\n",
    "    return r.text\n",
    "\n",
    "def calculateGC(seq):\n",
    "    '''\n",
    "    takes a genomic sequence and only counts AGCT items and returns simple GC content\n",
    "    Ambigious bases are ignored\n",
    "    '''\n",
    "    seq = seq.lower()\n",
    "    d = {}\n",
    "    for c in \"agct\":\n",
    "        d[c] = seq.count(c)\n",
    "    gc = (d['g'] + d['c'])/(d['a'] + d['g'] + d['c'] + d['t'])\n",
    "    return gc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "class gff3():\n",
    "    '''\n",
    "    This is a custom class for parsing esembles gff3 format\n",
    "    Usually you might want to resort to known established parsers such as \n",
    "    provided by biopython: https://biopython.org/wiki/GFF_Parsing\n",
    "    But their gff parser is not yet done. \n",
    "    For fasta and other file formats you can always rely on them though.\n",
    "    '''\n",
    "    def __init__(self, path, autoload = True):\n",
    "        '''\n",
    "        When initialized the class will always execute this __init__ function.\n",
    "        So whatever we do here, will always be done.\n",
    "        '''\n",
    "        self.path = path\n",
    "        if autoload:\n",
    "            self.load()\n",
    "    \n",
    "    def load(self):\n",
    "        '''\n",
    "        In this function we load the gff3 file by using pandas\n",
    "        '''\n",
    "        knownnames = ['seqid', 'source', 'type',\n",
    "                      'start', 'end', 'score', \n",
    "                      'strand', 'phase', 'attributes']\n",
    "        self.df = pd.read_csv(self.path, sep = '\\t', comment=\"#\", names = knownnames)\n",
    "    \n",
    "    def filter(self, column = \"type\", match = \"CDS\"):\n",
    "        '''\n",
    "        This filter function takes two arguments and returns a subset\n",
    "        based on these\n",
    "        '''\n",
    "        return self.df[self.df[column] == match]\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load the mouse gff3 file\n",
    "c = gff3('data/Mus_musculus.GRCm38.98.gff3.gz')\n",
    "# then filter for CDS reagions\n",
    "df = c.filter()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "# initialize an empty list to store the results\n",
    "res = []\n",
    "# iterate the dataframe after sampleing\n",
    "for index, row in df.sample(100).iterrows():\n",
    "    # fetch the sequence using the esemble API\n",
    "    seq = get_genomic_content(row.seqid, row.start, row.end, species = 'mouse')\n",
    "    # calculate the GC content\n",
    "    gc = calculateGC(seq)\n",
    "    # create a dictionary for each result\n",
    "    result = {'species': 'mouse',\n",
    "              'gc': gc}\n",
    "    # and save this in out result list\n",
    "    res.append(result)\n",
    "\n",
    "# transform the dictionary into a data frame\n",
    "result_df = pd.DataFrame(res)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[<matplotlib.axes._subplots.AxesSubplot object at 0x7f85a28949e8>]],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX4AAAEICAYAAABYoZ8gAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvnQurowAAF0VJREFUeJzt3X+QHGd95/H3BxuDS2tkGcNgZIUlF2HKeIODpmQqIWQWg5Flgvjh4uzzEQnsW0jhS3IlLnFCiFNAKs4dguLKrjgKVmxSiZcE8EVn+0xUDoviFAZWRrCywbEsNmWtfdJhyTILCmThmz+mNxmPZrWz3T3bMzyfV9XWdj/9dD/fR6397GzPj1ZEYGZm6XhW1QWYmdnycvCbmSXGwW9mlhgHv5lZYhz8ZmaJcfCbmSXGwW9mlhgHv5lZYhz8ZmaJcfCbLUDSqyR9TdJ3Jf21pE9L+ki2bZOkvZKelvSopA1V12vWLQe/WQeSTgPuAG4FzgJuB96abVsPfAr478CZwGuB6SrqNMtD/qwesxNJei3NsD83sh8SSfcBE8ALgO9HxH+rrkKz/PyI36yzFwMz8cxHRo9l39cAjy5/SWblcPCbdfYEsFqSWtrWZN8fA/7D8pdkVg4Hv1lnXwJ+BFwr6VRJm4D12bZbgHdJuljSsyStlvTyyio1WyIHv1kHEfFD4G3A1cBTwH8G7gR+EBFfAd4FfBw4BnwReElFpZotmZ/cNeuSpC8DN0fEn1Vdi1kRfsRvtgBJvyTpRdmlns3AzwL3VF2XWVGnVl2AWR87D/grYAVwALg8Ip6otiSz4nypx8wsMb7UY2aWmL681HP22WfH8PBwJWN/73vfY8WKFZWMXRbPoT8M+hwGvX5Iaw579uz5TkS8oJtj9mXwDw8PMzk5WcnYExMTNBqNSsYui+fQHwZ9DoNeP6Q1B0n/1O0xfanHzCwxDn4zs8Q4+M3MEuPgNzNLjIPfzCwxDn4zs8Q4+M3MEuPgNzNLjIPfzCwxffnOXbN+NXzdXV333Toyx5Yl9D+Z6RsuK+U4ZuBH/GZmyXHwm5klxsFvZpYYB7+ZWWIc/GZmiVk0+CWtkfQFSQ9JelDSr2ftZ0naJemR7PuqBfbfnPV5JLthtZmZVaibR/xzwNaIOB94NfA+SecD1wH3RsRa4N5s/RkknQVcD1wErAeuX+gXhJmZLY9Fgz8inoiIB7Ll7wLfBFYDm4Dbsm63AW/psPsbgV0RcSQijgK7gA1lFG5mZvks6Rq/pGHg54AvA7WIeCLb9P+AWoddVgOPtawfzNrMzKwiiojuOkpDwBeBP4iIz0l6KiLObNl+NCJWte3zfuC5EfGRbP2DwPGI+GiH448BYwC1Wm3d+Ph43jkVMjs7y9DQUCVjl8Vz6J2pmWNd962dDoeOlzPuyOqV5RxoCfr1HCxFSnMYHR3dExH1bo7Z1Uc2SHo28FngLyLic1nzIUnnRMQTks4BDnfYdQZotKyfC0x0GiMitgPbAer1elR1g+SUbs7cz/p1Dkv5CIatI3NsmyrnU1Gmr2qUcpyl6NdzsBSeQ2fdvKpHwC3ANyPiYy2bdgLzr9LZDPxNh90/D1wiaVX2pO4lWZuZmVWkm2v8vwC8E3idpL3Z10bgBuANkh4BXp+tI6ku6ZMAEXEE+DDw1ezrQ1mbmZlVZNG/QyPiPkALbL64Q/9J4JqW9R3AjrwFmplZufzOXTOzxDj4zcwS4+A3M0uMg9/MLDEOfjOzxDj4zcwS4+A3M0uMg9/MLDEOfjOzxDj4zcwS4+A3M0uMg9/MLDEOfjOzxDj4zcwS4+A3M0uMg9/MLDGL3ohF0g7gTcDhiLgga/s0cF7W5UzgqYi4sMO+08B3gR8Bc93eCNjMzHqnmztB3wrcCHxqviEi/uP8sqRtwLGT7D8aEd/JW6CZmZWrm1sv7pY03GlbdiP2dwCvK7csMzPrFUXE4p2awX/n/KWelvbXAh9b6BKOpG8DR4EA/iQitp9kjDFgDKBWq60bHx/vcgrlmp2dZWhoqJKxy+I59M7UzMn+uH2m2ulw6Hg5446sXlnOgZagX8/BUqQ0h9HR0T3dXk7v5lLPyVwJ3H6S7a+JiBlJLwR2SfpWROzu1DH7pbAdoF6vR6PRKFhaPhMTE1Q1dlk8h97Zct1dXffdOjLHtqmiP2JN01c1SjnOUvTrOVgKz6Gz3K/qkXQq8Dbg0wv1iYiZ7Pth4A5gfd7xzMysHEVezvl64FsRcbDTRkkrJJ0xvwxcAuwrMJ6ZmZVg0eCXdDvwJeA8SQclXZ1tuoK2yzySXizp7my1Btwn6evAV4C7IuKe8ko3M7M8unlVz5ULtG/p0PY4sDFbPgC8smB9ZmZWMr9z18wsMQ5+M7PEOPjNzBLj4DczS4yD38wsMQ5+M7PEOPjNzBLj4DczS4yD38wsMeV8dKCZ9dTwEj4VtCxbR+bYct1dTN9w2bKPbb3lR/xmZolx8JuZJcbBb2aWGAe/mVliHPxmZolx8JuZJaabO3DtkHRY0r6Wtt+XNCNpb/a1cYF9N0h6WNJ+SdeVWbiZmeXTzSP+W4ENHdo/HhEXZl93t2+UdApwE3ApcD5wpaTzixRrZmbFLRr8EbEbOJLj2OuB/RFxICJ+CIwDm3Icx8zMSqSIWLyTNAzcGREXZOu/D2wBngYmga0RcbRtn8uBDRFxTbb+TuCiiLh2gTHGgDGAWq22bnx8PNeEipqdnWVoaKiSscviOfTO1MyxrvvWTodDx3tYTI/N1z+yemXVpeTWr/+PlqLbOYyOju6JiHo3x8z7kQ1/DHwYiOz7NuDdOY8FQERsB7YD1Ov1aDQaRQ6X28TEBFWNXRbPoXe2LOGjE7aOzLFtanA/FWW+/umrGlWXklu//j9ail7MIdereiLiUET8KCJ+DPwpzcs67WaANS3r52ZtZmZWoVzBL+mcltW3Avs6dPsqsFbSSyWdBlwB7MwznpmZlWfRv0Ml3Q40gLMlHQSuBxqSLqR5qWcaeE/W98XAJyNiY0TMSboW+DxwCrAjIh7sySzMzKxriwZ/RFzZofmWBfo+DmxsWb8bOOGlnmZmVh2/c9fMLDEOfjOzxDj4zcwS4+A3M0vM4L67xJJWxT1ozX5S+BG/mVliHPxmZolx8JuZJcbBb2aWGAe/mVliHPxmZolx8JuZJcbBb2aWGAe/mVliHPxmZolZNPgl7ZB0WNK+lrb/Kelbkr4h6Q5JZy6w77SkKUl7JU2WWbiZmeXTzSP+W4ENbW27gAsi4meBfwR++yT7j0bEhd3e/d3MzHpr0eCPiN3Akba2v42IuWz1fpo3UjczswFQxjX+dwP/d4FtAfytpD2SxkoYy8zMClJELN5JGgbujIgL2to/ANSBt0WHA0laHREzkl5I8/LQf83+gug0xhgwBlCr1daNj48vcSrlmJ2dZWhoqJKxy5LCHKZmji1jNfnUTodDx6uuIr/5+kdWr6y6lNxS+FmYNzo6uqfbS+q5P49f0hbgTcDFnUIfICJmsu+HJd0BrAc6Bn9EbAe2A9Tr9Wg0GnlLK2RiYoKqxi5LCnPYMgCfx791ZI5tU4N7y4v5+qevalRdSm4p/CzkketSj6QNwG8Cb46I7y/QZ4WkM+aXgUuAfZ36mpnZ8unm5Zy3A18CzpN0UNLVwI3AGcCu7KWaN2d9Xyzp7mzXGnCfpK8DXwHuioh7ejILMzPr2qJ/h0bElR2ab1mg7+PAxmz5APDKQtWZWeWqus3l9A2XVTJuCvzOXTOzxDj4zcwS4+A3M0uMg9/MLDEOfjOzxDj4zcwS4+A3M0uMg9/MLDEOfjOzxDj4zcwS4+A3M0uMg9/MLDEOfjOzxDj4zcwS4+A3M0uMg9/MLDFdBb+kHZIOS9rX0naWpF2SHsm+r1pg381Zn0ckbS6rcDMzy6fbR/y3Ahva2q4D7o2ItcC92fozSDoLuB64iOaN1q9f6BeEmZktj66CPyJ2A0famjcBt2XLtwFv6bDrG4FdEXEkIo4CuzjxF4iZmS0jRUR3HaVh4M6IuCBbfyoizsyWBRydX2/Z5/3AcyPiI9n6B4HjEfHRDscfA8YAarXauvHx8bxzKmR2dpahoaFKxi5LCnOYmjm2jNXkUzsdDh2vuor8Br1+yD+HkdUryy8mp25/nkdHR/dERL2bYy56s/VuRERI6u43yMLH2A5sB6jX69FoNMoobckmJiaoauyypDCHLRXdAHwpto7MsW2qlB+xSgx6/ZB/DtNXNcovJqde/DwXeVXPIUnnAGTfD3foMwOsaVk/N2szM7OKFAn+ncD8q3Q2A3/Toc/ngUskrcqe1L0kazMzs4p0+3LO24EvAedJOijpauAG4A2SHgFen60jqS7pkwARcQT4MPDV7OtDWZuZmVWkq4tfEXHlApsu7tB3ErimZX0HsCNXdWZmVjq/c9fMLDEOfjOzxDj4zcwS4+A3M0uMg9/MLDEOfjOzxDj4zcwS4+A3M0uMg9/MLDEOfjOzxDj4zcwS4+A3M0uMg9/MLDEOfjOzxAz2fdUMgOG22xBuHZlbllsTTt9wWc/HMLPy+RG/mVlicge/pPMk7W35elrSb7T1aUg61tLn94qXbGZmReS+1BMRDwMXAkg6heZN1O/o0PXvI+JNeccxM7NylXWp52Lg0Yj4p5KOZ2ZmPaKIKH4QaQfwQETc2NbeAD4LHAQeB94fEQ8ucIwxYAygVqutGx8fL1xXHrOzswwNDVUydl5TM8eesV47HQ4d7/24I6tX9uzYi52H9jn3o+U6D70y6PVD/jn08v/2UnWbSaOjo3siot7NMQsHv6TTaIb6KyLiUNu25wE/johZSRuBT0TE2sWOWa/XY3JyslBdeU1MTNBoNCoZO69Or+rZNtX7F2z18lU9i52H9jn3o+U6D70y6PVD/jn00yvWus0kSV0HfxmXei6l+Wj/UPuGiHg6Imaz5buBZ0s6u4QxzcwspzKC/0rg9k4bJL1IkrLl9dl4T5YwppmZ5VTo7zhJK4A3AO9paXsvQETcDFwO/KqkOeA4cEWU8aSCmZnlVij4I+J7wPPb2m5uWb4RuLF9P/vJ0Mvr7Mv17mOzFPmdu2ZmiXHwm5klxsFvZpYYB7+ZWWIc/GZmiXHwm5klxsFvZpYYB7+ZWWIc/GZmiRnsj94zM+uBqj79dbk+FdSP+M3MEuPgNzNLjIPfzCwxDn4zs8Q4+M3MElM4+CVNS5qStFfSCTfKVdP/krRf0jckvaromGZmll9ZL+ccjYjvLLDtUmBt9nUR8MfZdzMzq8ByXOrZBHwqmu4HzpR0zjKMa2ZmHajoLXAlfRs4CgTwJxGxvW37ncANEXFftn4v8FsRMdnWbwwYA6jVauvGx8cL1ZXX7OwsQ0NDlYyd19TMsWes106HQ8crKqYknkP1Br1+GLw5jKxeeUJbt5k0Ojq6JyLq3YxTxqWe10TEjKQXArskfSsidi/1INkvjO0A9Xo9Go1GCaUt3cTEBFWNnVf7vWm3jsyxbWqw35TtOVRv0OuHwZvD9FWNE9p6kUmFL/VExEz2/TBwB7C+rcsMsKZl/dyszczMKlAo+CWtkHTG/DJwCbCvrdtO4FeyV/e8GjgWEU8UGdfMzPIr+jdQDbhD0vyx/jIi7pH0XoCIuBm4G9gI7Ae+D7yr4JhmZlZAoeCPiAPAKzu039yyHMD7ioxjZmbl8Tt3zcwS4+A3M0uMg9/MLDEOfjOzxDj4zcwS4+A3M0uMg9/MLDEOfjOzxDj4zcwS4+A3M0uMg9/MLDEOfjOzxDj4zcwS4+A3M0vM4NyTrEvDbbchXKqtI3Mn3MrQzOwniR/xm5klJnfwS1oj6QuSHpL0oKRf79CnIemYpL3Z1+8VK9fMzIoqcqlnDtgaEQ9k993dI2lXRDzU1u/vI+JNBcYxM7MS5X7EHxFPRMQD2fJ3gW8Cq8sqzMzMekPNW+IWPIg0DOwGLoiIp1vaG8BngYPA48D7I+LBBY4xBowB1Gq1dePj47lqmZo5lmu/ebXT4dDxQoeonOfQHwZ9DoNePwzeHEZWrzyhbXZ2lqGhoUX3HR0d3RMR9W7GKRz8koaALwJ/EBGfa9v2PODHETEraSPwiYhYu9gx6/V6TE5O5qqnjFf1bJsa7Bc7eQ79YdDnMOj1w+DNYfqGy05om5iYoNFoLLqvpK6Dv9CreiQ9m+Yj+r9oD32AiHg6Imaz5buBZ0s6u8iYZmZWTJFX9Qi4BfhmRHxsgT4vyvohaX023pN5xzQzs+KK/A30C8A7gSlJe7O23wF+CiAibgYuB35V0hxwHLgiynhSwczMcssd/BFxH6BF+twI3Jh3DDMzK5/fuWtmlhgHv5lZYhz8ZmaJcfCbmSXGwW9mlhgHv5lZYhz8ZmaJcfCbmSXGwW9mlhgHv5lZYhz8ZmaJcfCbmSXGwW9mlhgHv5lZYhz8ZmaJcfCbmSWm6D13N0h6WNJ+Sdd12P4cSZ/Otn9Z0nCR8czMrLgi99w9BbgJuBQ4H7hS0vlt3a4GjkbEzwAfB/4o73hmZlaOIo/41wP7I+JARPwQGAc2tfXZBNyWLX8GuHj+5utmZlYN5b33uaTLgQ0RcU22/k7gooi4tqXPvqzPwWz90azPdzocbwwYy1bPAx7OVVhxZwMn1DdgPIf+MOhzGPT6Ia05vCQiXtDNAXPfbL1sEbEd2F51HZImI6JedR1FeA79YdDnMOj1g+ewkCKXemaANS3r52ZtHftIOhVYCTxZYEwzMyuoSPB/FVgr6aWSTgOuAHa29dkJbM6WLwf+LvJeWzIzs1LkvtQTEXOSrgU+D5wC7IiIByV9CJiMiJ3ALcCfS9oPHKH5y6HfVX65qQSeQ38Y9DkMev3gOXSU+8ldMzMbTH7nrplZYhz8ZmaJSSr4u/iIifdKmpK0V9J98+9EljQs6XjWvlfSzctf/b/VeNI5tPR7u6SQVG9p++1sv4clvXF5Kj6hrlz1D9I5kLRF0v9vqfWalm2bJT2SfW1u33e5FJzDj1ra21/QsWy6+b8k6R2SHpL0oKS/bGkfiPOQ9VloDvnPQ0Qk8UXzCehHgZ8GTgO+Dpzf1ud5LctvBu7JloeBfYMwh6zfGcBu4H6gnrWdn/V/DvDS7DinDFD9A3MOgC3AjR32PQs4kH1flS2vGqQ5ZNtmB+Q8rAW+Nv9vDLxwAM9DxzkUPQ8pPeJf9CMmIuLpltUVQL89893Nx2QAfJjm5yL9c0vbJmA8In4QEd8G9mfHW05F6u8X3c6hkzcCuyLiSEQcBXYBG3pU58kUmUO/6GYO/wW4Kfu3JiIOZ+2DdB4WmkMhKQX/auCxlvWDWdszSHpf9tES/wP4tZZNL5X0NUlflPSLvS11QYvOQdKrgDURcddS910GReqHATkHmbdL+oakz0iaf6NjP5yDpdTRaQ4Az5U0Kel+SW/paaUL62YOLwNeJukfslo3LGHf5VBkDlDgPPTNRzb0i4i4CbhJ0n8CfpfmG9CeAH4qIp6UtA7435Je0fYXQuUkPQv4GM0/0wfOIvUPxDnI/B/g9oj4gaT30PygwtdVXNNSnWwOL4mIGUk/DfydpKmIeLSyShd2Ks1LJQ2anyywW9JIpRUtXcc5RMRTFDgPKT3i7+YjJlqNA28ByC6PPJkt76F5Xe5lParzZBabwxnABcCEpGng1cDO7AnSpc6/F3LXP0DngIh4MiJ+kK1+EljX7b7LpMgciIiZ7PsBYAL4uV4Wu4Bu/i0PAjsj4l+yy5v/SDNEB+Y8sPAcip2H5X5Co6ovmr85D9B8YnP+iZRXtPVZ27L8yzTfgQzwArInQmk+ETMDnNWPc2jrP8G/Pzn6Cp755O4Blv/J3SL1D8w5AM5pWX4rcH+2fBbwbZpPKK7KlgdtDquA52TLZwOP0OEJ+j6ZwwbgtpZaHwOeP2DnYaE5FDoPyzrRqr+AjTR/Yz4KfCBr+xDw5mz5E8CDwF7gC/MnAXh7S/sDwC/36xza+v5bcGbrH8j2exi4dJDqH6RzAPxhVuvXs/9HL2/Z9900n1jfD7xr0OYA/DwwlbVPAVf38RxE89LhQ1mtVwzgeeg4h6LnwR/ZYGaWmJSu8ZuZGQ5+M7PkOPjNzBLj4DczS4yD38wsMQ5+M7PEOPjNzBLzr1aAW8Lpj//IAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# plot simple historgram\n",
    "result_df.hist()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that you saw how to do it for one species, you can do the same for human and zebrafish and plot them together. You can merge pandas dataframes using `pd.concat`. Check the documentary for help. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "files = ['Danio_rerio.GRCz11.98.gff3.gz', 'Homo_sapiens.GRCh38.98.gff3.gz',  'Mus_musculus.GRCm38.98.gff3.gz']\n",
    "species =  ['Zebrafish', 'Human', 'Mouse']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "only scroll down If you want to see my solution. Else work in here\n"
     ]
    }
   ],
   "source": [
    "print(\"only scroll down If you want to see my solution. Else work in here\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Solution\n",
    "class gffsample(gff3):\n",
    "    '''\n",
    "    this class inherits the functionality of the gff class and \n",
    "    expands it with new functionality for sampeling. But relies on outside functions\n",
    "    for computing gc and fetching the data.\n",
    "    Would be nice to include these into this class.\n",
    "    '''\n",
    "    def __init__(self, path, species = \"human\"):\n",
    "        '''\n",
    "        When initialized the class will always execute this __init__ function.\n",
    "        So whatever we do here, will always be done.\n",
    "        '''\n",
    "        self.path = path\n",
    "        # filter\n",
    "        self.load()\n",
    "        # also filter this time\n",
    "        self.df = self.filter()\n",
    "        # set the species\n",
    "        self.species = species\n",
    "        self.df_gc_created = False\n",
    "    \n",
    "    def sample(self, n = 100):\n",
    "        self.sampled = self.df.sample(n)\n",
    "    \n",
    "    def make_GC(self):\n",
    "        res = []\n",
    "        for index, row in self.sampled.iterrows():\n",
    "            # fetch the sequence using the esemble API\n",
    "            seq = get_genomic_content(row.seqid, row.start, row.end, species = self.species)\n",
    "            # calculate the GC content\n",
    "            gc = calculateGC(seq)\n",
    "            # create a dictionary for each result\n",
    "            result = {'species': self.species,\n",
    "                      'gc': gc}\n",
    "            # and save this in out result list\n",
    "            res.append(result)\n",
    "        # transform the dictionary into a data frame\n",
    "        self.df_gc = pd.DataFrame(res)\n",
    "    \n",
    "    def get_GC(self):\n",
    "        # if not done already compute the GC content by querying the esemble api\n",
    "        if self.df_gc_created == False:\n",
    "            self.make_GC()\n",
    "            self.df_gc_created = True\n",
    "        \n",
    "        return self.df_gc\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Samepling for Danio_rerio.GRCz11.98.gff3.gz of Zebrafish\n",
      "Samepling for Homo_sapiens.GRCh38.98.gff3.gz of Human\n",
      "Samepling for Mus_musculus.GRCm38.98.gff3.gz of Mouse\n"
     ]
    }
   ],
   "source": [
    "dataframes = []\n",
    "for file, specie in zip(files, species):\n",
    "    print(f'Sampeling file: {file} ({specie})')\n",
    "    g = gffsample(f'data/{file}', species = specie)\n",
    "    g.sample(100)\n",
    "    dataframes.append(g.get_GC())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[<matplotlib.axes._subplots.AxesSubplot object at 0x7f84f8078dd8>,\n",
       "        <matplotlib.axes._subplots.AxesSubplot object at 0x7f84f95e6080>],\n",
       "       [<matplotlib.axes._subplots.AxesSubplot object at 0x7f84df18b2e8>,\n",
       "        <matplotlib.axes._subplots.AxesSubplot object at 0x7f84e12f3b70>]],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX8AAAENCAYAAADnrmWtAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvnQurowAAGOBJREFUeJzt3Xu0nXV95/H3J5FUR+7NMaZAOFgCCu3IJVIcp5WLILeu0E5hQIdGZRntlLY4tENkOp22aie4xgutWhsLi1gtl4piRphSVooX5CIJppoQFY2JJA0kgWQC4iBJPvPH85ywOZ6Ts8/Zl2fv/Xxea+11nuv+ffc++/fdv/17Lj/ZJiIi6mVa1QFERET3JflHRNRQkn9ERA0l+UdE1FCSf0REDSX5R0TUUJJ/REQNJfm3gaT1kt40atnbJN1bVUwRVSvrxU8lzRy1/JuSLGm4msgCkvwjorN+CFw6MiPpl4F/U104MSLJvwvKVs7RDfM3Snp/OX2apI2S/qukLZI2S7pQ0nmSvifpKUnXNOx7iqT7Je0ot/2YpBmjynq3pEfLbT4uSd19xRF7/R3w2w3zC4BPj8xIOkjSpyVtlbRB0h9Lmlau+1NJn2nYdrj8fL+knH+bpHWSnpb0Q0lvbdj2HZLWStou6S5JR3b8lfaZJP/e8ErgpcBhwJ8AnwL+E3Ay8KvAf5d0VLntbuA9wEzg9cCZwH8e9XwXAK8D/i1wMfDmDscfMZ4HgAMlvUbSdOAS4DMN6/8KOAh4FfBGii+Kt0/0pJJeDvwlcK7tA4B/B6wq180HrgF+ExgCvgbc1K4XNCiS/Nvn9rKlvUPSDuATk9j3eeADtp8HbqZI7NfZftr2GuAR4LUAtlfafsD2Ltvrgb+hqDSNFtveYftHwD3ACa29tIiWjLT+zwLWApvK5SNfBu8tP+vrgQ8BlzX5vHuAX5L0Mtuby7oC8G7gf9pea3sX8BfACWn9v1iSf/tcaPvgkQc/2xrflydt7y6nf1L+faJh/U+A/QEkHSPpS5Iel7ST4oP9ogNqwOMN08+O7BtRkb8D3gK8jYYuH4rP7X7AhoZlGyh+Ae+T7R8D/5Ei0W+WdIekV5erjwSua2iIPQWomeetkyT/7niWFx/kemULz/XXwHeAubYPpPh5mz796Fm2N1Ac+D0P+HzDqm0Uv3obW+RzeOGXwY/ZR72xfZfts4DZFHXiU+Wqx4B3NTbGbL/M9n3tek2DIMm/O1YBb5E0XdI5/Gw3zWQcAOwEnilbOr/TjgAjOuxy4IyyxT5iN3Ar8AFJB5TdMv+FF44JrAJ+TdIcSQcB7x3ZUdIsSfPLvv/ngGcouoEAPgm8V9Lx5bYHSbqoky+uHyX5d8cfAL8O7ADeCtzewnP9IcVP6KcpWjq3tBxdRIfZ/oHtFWOs+j2KFv464F7g74Ebyn3upvh8fwtYCXypYb9pFF8U/0rRrfNGyoaQ7S8A1wI3l12jq4Fz2/+q+psymEtERP2k5R8RUUNJ/hERNZTkHxFRQ0n+ERE1lOQfEVFDL+lmYTNnzvTw8HA3i4waWLly5TbbQ1XHMRmpC9EpzdaHrib/4eFhVqwY61TfiKmTtGHirXpL6kJ0SrP1Id0+ERE1lOQfEVFDXe326RXDi+7oWlnrF5/ftbIiBtFU62vq3r6l5R8RUUNJ/hERNZTkHxFRQ0n+ERE1VMsDvt2Ug1UR0YsmbPlLukHSFkmrG5b9qaRNklaVj/M6G2ZERLRTM90+NwLnjLH8I7ZPKB93tjesiIjopAmTv+2vUgyTFhERA6KVA75XSPpW2S10SNsiioiIjpvqAd+/Bt4HuPz7IeAdY20oaSGwEGDOnDlTLC6aNZUDzDm43BxJRwCfBmZRfPaX2L5O0qEUA40PA+uBi21vryrOiGZMqeVv+wnbu23vAT4FnLKPbZfYnmd73tBQX911N2K0XcBVto8DTgV+V9JxwCJgue25wPJyPqKnTSn5S5rdMPsbwOrxto0YFLY32364nH4aWAscBswHlpabLQUurCbCiOZN2O0j6SbgNGCmpI3A/wBOk3QCxU/f9cC7OhhjRM+RNAycCDwIzLK9uVz1OEW3UERPmzD52750jMXXdyCWiL4gaX/gNuBK2zsl7V1n25I8zn45/hU9I1f49qhu3nY6midpP4rE/1nbny8XPyFptu3NZZfolrH2tb0EWAIwb968Mb8gIrol9/aJaJKKJv71wFrbH25YtQxYUE4vAL7Y7dgiJist/4jmvQG4DPi2pFXlsmuAxcCtki4HNgAXVxRfRNOS/COaZPteQOOsPrObsUS0Kt0+ERE1lOQfEVFDSf4RETWU5B8RUUNJ/hERNZTkHxFRQ0n+ERE1lOQfEVFDucgrIiZtqveeysBBvSMt/4iIGkryj4iooQmTfzlA+xZJqxuWHSrpbkmPln8zgHtERB9pps//RuBjFANXjxgZs3SxpEXl/NXtDy8iBknGqegdE7b8bX8VeGrU4oxZGhHRx6ba558xSyMi+ljLB3xtm2Ig9zFJWihphaQVW7dubbW4iIhog6km/yfKsUrZ15ilUIxbanue7XlDQ0NTLC4iItppqsk/Y5ZGRPSxZk71vAm4HzhW0sZynNLFwFmSHgXeVM5HRESfmPBUT9uXjrMqY5ZGRM/KLSj2LVf4RkTUUJJ/REQN5a6eETWWK27rKy3/iIgaSvKPiKihnun2mcrPz7oclY/eIekG4AJgi+1fKpcdCtwCDAPrgYttb68qxohmpOUfMTk3AueMWjZyl9u5wPJyPqKnJflHTELuchuDIsk/onW5y230nZ7p85+KnKYWvca2JY15l1tJC4GFAHPmzOlqXBGjpeUf0bqm7nKbO9xGL0nyj2hd7nIbfSfJP2IScpfbGBR93ecf0W25y20MirT8IyJqqKWWv6T1wNPAbmCX7XntCCoiIjqrHd0+p9ve1obniYgW5NTnmIx0+0RE1FCryd/AP0laWV7AEhERfaDVbp9/b3uTpFcAd0v6Tnnvk71yVWNERO9pqeVve1P5dwvwBeCUMbbJVY0RET1myslf0sslHTAyDZwNrG5XYBER0TmtdPvMAr4gaeR5/t72P7YlqoiI6KgpJ3/b64DXtjGWiIjokpzqGRFRQ0n+ERE1lBu7RUQ0mMqV0usXn9+BSDorLf+IiBpK8o+IqKF0+8SUTfVGYv34Ezli0CT5R0S0qB8bQun2iYiooST/iIgaSvKPiKih9PlHRoCKqKG0/CMiaigt/4iIilR5llBa/hERNZSWf3RdXe6dEtHL0vKPiKihllr+ks4BrgOmA39re3FbooroQ+2qDzn7KrqhlTF8pwMfB84FjgMulXRcuwKL6CepD9FvWun2OQX4vu11tn8K3AzMb09YEX0n9SH6SivdPocBjzXMbwR+ZfRGkhYCC8vZZyR9t4UyJ2MmsK1LZaXsDpera/e5+thOlTsJE9aHLteFKj+DY0k8+zapeCaoD0c28xwdP9vH9hJgSafLGU3SCtvzul1uXcuu+jVXUe5kdbMuVPn/GEvi2bcq4mml22cTcETD/OHlsog6Sn2IvtJK8n8ImCvpKEkzgEuAZe0JK6LvpD5EX5lyt4/tXZKuAO6iOLXtBttr2hZZ67re1VTzsuv4mvfqwfpQ+XsySuLZt+53jdvudpkREVGxXOEbEVFDSf4RETWU5B8RUUMDm/wlnVR1DN0k6UBJJ0s6pOpYuk3SzKpj6DV1+/xPRp3rSqOBSP6SThr1OBlYJunETlcCSe9omD5c0nJJOyTdJ+mYDpb7mZGkJ+nNwGrgWmCVpIs6VW5Z3lOS/lbSmZLUybLGKPtcST+UdG/5/10DPChpo6QzuxlLr6jy8z9OPJXUiX3EU1ldGSeeyurPi9ju+wewB7gPuKfh8ZPy7z93uOyHG6Zvpbh8fxrwG8DyDpb77Ybp+4Dhcnom8C8dfs3fBa4Avk5xIdN1wKld+l+vAl4DvB54cqTcctnD3Yih1x5Vfv7HiaeSOrGPeCqrK+PEU1n9aXwMRMsfuAh4Hvig7dNtnw48Xk6f0cU4jrG9xPYe218ADu1gWdMkHVhO7wF+BGB7G52/bcePbX/M9hsokvAm4BOS1kn6iw6Xvcf2Wtv3A8/afgDA9loG5JfsFPTK538s3awT46myroylyvqz10BUFtu3AecDZ0v6B0lzgG5dwHC4pL+U9FfAkKT9GtbtN95ObfBnwD3lT+yvA/8gaYGkG4F/7GC5AHt/qtr+ke0P2j4JOA94rsNl75D0Lkl/BGyX9B5Jh0laADzT4bJ7UsWf/7FUVSfGU2VdGUuV9WevgRnG0fYzwHsknQgsBfbvUtF/1DC9oix3u6RX0sHL+23fKulh4J3AMRT/y1OBm2zf1alyS/eME9N3KCpaJy0A/piiBXc2cCnFVbUbKN6LWqrw8z+WSurEeCquK2Opsv7sNZBX+JYHUQ6wvbPqWCK6LZ//aMZAdPuM5sJOAEkXVBVHVWXX8TVXXXYv6ZXP/1gSz751M56BTP6jvK6GZdfxNVdddq/qtfck8exb1+IZmG4fSa+mGDbvsHLRJmBZeRbIQJZdx9dcddm9qtfek8TT+/EMRMtf0tUUY6YK+Eb5EHCTpEWDWHYdX3PVZfeqXntPEk9/xDMQLX9J3wOOt/38qOUzgDW25w5a2XV8zVWX3at67T1JPP0Rz0C0/ClO+/uFMZbPLtcNYtl1fM1Vl92reu09STz71hPxDMp5/lcCyyU9CjxWLpsDHE1xGfUgll3H11x12b2q196TxNMH8QxEtw+ApGnAKbz4AMpDtncPatl1fM1Vl92reu09STy9H8/AJP+IiGjeoPT5R0TEJCT5R0TUUJJ/REQNJfl3gaS3Sbq3zc/5O5KekPSMpJ8v/75qgn1Ok7SxnXFERH9K8m+SpLeWCXb0w5L+pMux7Ad8GDjb9v62nyz/rutmHBHRv5L8m2T7s2WC3fugOF/3CeBT7SpHhYn+L7OAlwJr2lVuRNRLkv8UlYNmfBS4xPZmSQdJul7SZkmbJL1f0vQX76KPSfq/kr6jhsHGJX1Z0gckfR14FniVpLdLWivpaRXDu72r3PYYijFAoRjV6p/L5ZZ0dDl9nqRHyn03SfrDUbFfJWlLGevbO/YmRUTPGpQrfLtK0sHA54D32f5yufhGYAvFVXovB75EcfXe35Trf6XcZybwm8DnJR1l+6ly/WXAuRSJXcCxwAXAOuDXgP8j6SHbD0s6HvghcLDtXWOEeD1wse2vSToEOKph3SuBgyguLjkL+Jyk221vb+EtiYg+k5b/JEkS8GlgNfDBctksivE3r7T9Y9tbgI8AlzTsugX4qO3nbd9CkeTPb1h/o+01tneV29xh+wflwBxfAf4J+NUmw3weOE7Sgba323541Lo/L8u4k2Lc22Mn+z5ERH9L8p+8q4HjgQV+4fLoIykGpt4saYekHRQt/lc07LepYXsoxpxtvLnTYw3TSDpX0gOSniqf7zyKXw3N+A/l9hskfUXS6xvWPTnq18KzVDvea0RUIMl/EiSdBvw34Lds72hY9RjwHDDT9sHl40Dbxzdsc1j5q2HEHOBfG+b3fjFI+jngNuB/AbNsHwzcSdEdNCHbD9meT/Hlcztwa7OvMSLqIcm/SZJmUwzAcKXtbzaus72ZolvmQ5IOlDRN0i9KemPDZq8Afl/SfpIuAl5DkdDHMgP4OWArsEvSucDZTcY5ozwt9aDyfuE7qe+tjiNiHEn+zXsnxSmW141xrv8ngd+mSNqPANspDu7Obtj/QWAusA34AMWvhyfHKsj208DvU7TYtwNvAZZNItbLgPWSdgLvBt46iX0jogZyV8+IiBpKyz8iooaS/CMiaijJPyKihpL8IyJqKMk/IqKGunpvn5kzZ3p4eLibRUYNrFy5cpvtoarjiOgnXU3+w8PDrFixoptFRg1I2lB1DBH9Jt0+ERE1lOQfEVFDuZ9/MLzojq6Wt37x+RNvFBEdNWHLX9JLJX1D0r9IWiPpz8rlR0l6UNL3Jd0iaUbnw42IiHZoptvnOeAM268FTgDOkXQqcC3wEdtHU9x87PLOhRkREe00YfIvR5J6ppzdr3wYOIPizpUAS4ELOxJhRES0XVMHfCVNl7SKYijCu4EfADsaRoTaSDEmbERE9IGmkr/t3bZPAA4HTgFe3WwBkhZKWiFpxdatW6cYZkREtNOkTvUshy68B3g9cLCkkbOFDgc2jbPPEtvzbM8bGspFmBERvaCZs32GJB1cTr8MOAtYS/El8FvlZguAL3YqyIiIaK9mzvOfDSyVNJ3iy+JW21+S9Ahws6T3A98Eru9gnBER0UYTJn/b3wJOHGP5Oor+/4iI6DO5vUNERA0l+UdE1FCSf0REDSX5R0TUUJJ/REQNJflHRNRQkn9ERA0l+UdE1FCSf0REDSX5R0TUUJJ/REQNJflHRNRQkn9ERA0l+UdE1FCSf0REDSX5R0TUUJJ/REQNNTOG7xGS7pH0iKQ1kv6gXH6opLslPVr+PaTz4UZERDs00/LfBVxl+zjgVOB3JR0HLAKW254LLC/nIyKiD0yY/G1vtv1wOf00sBY4DJgPLC03Wwpc2KkgIyKivSbV5y9pmGIw9weBWbY3l6seB2aNs89CSSskrdi6dWsLoUZERLs0nfwl7Q/cBlxpe2fjOtsGPNZ+tpfYnmd73tDQUEvBRkREezSV/CXtR5H4P2v78+XiJyTNLtfPBrZ0JsSIiGi3Zs72EXA9sNb2hxtWLQMWlNMLgC+2P7yIiOiElzSxzRuAy4BvS1pVLrsGWAzcKulyYANwcWdCjIiIdpsw+du+F9A4q89sbzgREdENucI3IqKGkvwjImooyT8iooaS/CMiaijJPyKihpo51TP6yPCiO6oOYUJTiXH94vM7EElEfaXlHxFRQ0n+ERE1lOQfEVFDSf4RETWU5B8RUUNJ/hERNZTkHxFRQ0n+ERE1lOQfEVFDSf4RETXUzDCON0jaIml1w7JDJd0t6dHy7yGdDTMiItqpmZb/jcA5o5YtApbbngssL+cjIqJPTJj8bX8VeGrU4vnA0nJ6KXBhm+OKiIgOmmqf/yzbm8vpx4FZbYonIiK6oOUDvrYNeLz1khZKWiFpxdatW1stLiIi2mCqyf8JSbMByr9bxtvQ9hLb82zPGxoammJxERHRTlNN/suABeX0AuCL7QknIiK6oZlTPW8C7geOlbRR0uXAYuAsSY8CbyrnIyKiT0w4jKPtS8dZdWabY4mIiC7JFb4RETWU5B8RUUNJ/hERNZTkHxFRQ0n+ERE1lOQfEVFDSf4RETWU5B8RUUNJ/hERNZTkHxFRQ0n+ERE1lOQfEVFDSf4RETWU5B8RUUNJ/hERNZTkHxFRQ0n+ERE11FLyl3SOpO9K+r6kRe0KKiIiOmvCYRzHI2k68HHgLGAj8JCkZbYfmcrzDS+6Y9L7rF98/lSKmrKpxBgR0YtaafmfAnzf9jrbPwVuBua3J6yIiOikVpL/YcBjDfMby2UREdHjptzt0yxJC4GF5ewzkr47zqYzgW2Teu5rW4msZZOOt2L9FO/PxDrB//rITgYTMYhaSf6bgCMa5g8vl72I7SXAkomeTNIK2/NaiKerEm/n9FOsEf2qlW6fh4C5ko6SNAO4BFjWnrAiIqKTptzyt71L0hXAXcB04Abba9oWWUREdExLff627wTubFMsE3YN9ZjE2zn9FGtEX5LtqmOIiIguy+0dIiJqKMk/IqKGkvwjImqo0uQv6VBJh1YZQx1IOqnqGJol6UBJJ0s6pOpYIgZZ15O/pDmSbpa0FXgQ+IakLeWy4W7HMxFJR5SxfU3SNZL2a1h3e5WxjUXSSaMeJwPLJJ3Yi18Ckj4jaWY5/WZgNXAtsErSRZUGFzHAun62j6T7gY8Cn7O9u1w2HbgIuNL2qV0NaAKS7gZuAx4ALgdOBn7d9pOSvmn7xEoDHEXSHopYn2tYfGq5zLbPqCSwcUj6tu1fLqfvA95ie335hbDc9murjTBiMFXR7TPT9i0jiR/A9m7bNwM/X0E8Exmy/Unbq2z/HvAJ4KuSfhHoxfNkLwKeBz5o+3TbpwOPl9M9lfhL0yQdWE7vAX4EYHsbXbj3VERdVdHyvxl4CljKC3cFPQJYQPHFcHFXA5qApDXAybb/X8OyNwGfBF5ue3ZlwY1D0v7A+yjut3QV8GXbr6o2qrFJuhi4mmJsiGOBoyluE3I68KTtqyoML2JgVZH8Z1B0n8znhVtAbwT+N3C97efG27cKkt4DPGz7K6OWn0jRuj6rmsgmVsb4YeB426+oOp7xSDoaeCdwDEVrfyNwu+27Kg0sYoDlCt8BJ0nAAbZ3Vh1LRPSOnjrPX9IFVccwGf0Qrws7oT/ibdRv8Ub0k55K/sDrqg5gkhJvZ/VbvBF9o5JuH0mv5sV9/puAZbbXdj2YJiTezuq3eCMGQRUXeV1NMdi7gG+UDwE3SVrU7Xgmkng7q9/ijRgUVZzt8z2Ks0+eH7V8BrDG9tyuBjSBxNtZ/RZvxKCoos9/D/ALYyyfXa7rNYm3s/ot3oiBUMUVlFcCyyU9ygsXec2huLjnigrimUji7ax+izdiIFR1wHcacAovPsD3UOMtH3pJ4u2sfos3YhDkIq+IiBrqtfP8IyKiC5L8IyJqKMk/IqKGkvwjImooyT8ioob+P4KF3Bz1mom8AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 4 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "pd.concat(dataframes).hist(by = \"species\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
